File: crates/addr-spec/0.5.2/src/ascii.c
#include "ascii.h"

#define true 1
#define false 0

typedef int bool;

typedef struct {
  const char *first;
  const char *last;
} String;

// NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
bool has_chr(String str, char chr) {
  for (; str.first < str.last; ++str.first) {
    if (*str.first == chr) {
      return true;
    }
  }
  return false;
}

// NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
size_t ascii_escape_len(char esc_chr, const char *src, size_t len,
                        const char *cntl_set, size_t set_len) {
  String str = {cntl_set, cntl_set + set_len};
  size_t escape_count = 0;
  const char *end = src + len;
  for (; src < end; ++src) {
    char chr = *src;
    if (chr == esc_chr || has_chr(str, chr)) {
      escape_count += 1;
    }
  }
  return len + escape_count;
}

void ascii_escape(char esc_chr, const char *src, size_t len, char *dst,
                  const char *cntl_set, size_t set_len) {
  String str = {cntl_set, cntl_set + set_len};
  const char *end = src + len;
  for (; src < end; ++src) {
    char chr = *src;
    if (chr == esc_chr || has_chr(str, chr)) {
      *dst++ = esc_chr;
    }
    *dst++ = chr;
  }
}

File: crates/addr-spec/0.5.2/src/ascii.rs
use std::os::raw::c_char;

include!(concat!(env!("OUT_DIR"), "/ascii.rs"));

pub fn escape(esc_chr: char, src: &str, cntl_set: &str) -> String {
    let len = unsafe {
        ascii_escape_len(
            esc_chr as c_char,
            src.as_ptr() as *const c_char,
            src.len(),
            cntl_set.as_ptr() as *const c_char,
            cntl_set.len(),
        )
    };
    let mut dst = Vec::with_capacity(len);
    unsafe {
        ascii_escape(
            esc_chr as c_char,
            src.as_ptr() as *const c_char,
            src.len(),
            dst.as_mut_ptr() as *mut c_char,
            cntl_set.as_ptr() as *const c_char,
            cntl_set.len(),
        );
        dst.set_len(len);
        String::from_utf8_unchecked(dst)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_escape() {
        assert_eq!(escape('\\', "", "\""), "");
        assert_eq!(escape('\\', "abc", "\""), "abc");
        assert_eq!(escape('\\', "a\\b", "\""), "a\\\\b");
        assert_eq!(escape('\\', "a\"b", "\""), "a\\\"b");
        assert_eq!(escape('\\', "a\\\"b", "\""), "a\\\\\\\"b");
        assert_eq!(escape('\\', "😄\"😄😄", "\""), "😄\\\"😄😄");
    }
}
